[WIP] 🐛 clusterctl upgrade: ensure old controllers are gone before installing the new ones

sbueringer opened the pull request:
Signed-off-by: Stefan Büringer buringerst@vmware.com

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
PR is WIP. There are probably better way to do this and I'll talk to @fabriziopandini about them tomorrow :)
I mostly pushed it to run the tests and find out if it would fix the issue.


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5349

sbueringer commented:
/test pull-cluster-api-e2e-full-main

sbueringer commented:
First run was successful (I didn't wait until the end: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/5354/pull-cluster-api-e2e-full-main/1442921166653100032)

/test pull-cluster-api-e2e-full-main

fabriziopandini commented on cmd/clusterctl/client/cluster/components.go:
```diff
@@ -209,6 +215,12 @@ func (p *providerComponents) Delete(options DeleteOptions) error {
 		}
 
+		if obj.GetKind() == deploymentKind {
+			if err := scaleDownDeployment(ctx, cs, obj); err != nil {
```
Let's document why we are doing this

sbueringer replied:
Absolutely, this wasn't meant to be ready for review yet :)

fabriziopandini commented on cmd/clusterctl/client/cluster/components.go:
```diff
@@ -209,6 +215,12 @@ func (p *providerComponents) Delete(options DeleteOptions) error {
 
+		if obj.GetKind() == deploymentKind {
+			if err := scaleDownDeployment(ctx, cs, obj); err != nil {
+				return err
```
```suggestion
				return errors.Wrap(err, "failed to scale down deployment")
```

sbueringer replied:
I added it one layer below as I had to do that anyway

fabriziopandini commented on cmd/clusterctl/client/cluster/components.go:
```diff
@@ -224,6 +236,54 @@ func (p *providerComponents) Delete(options DeleteOptions) error {
+		return err
+	}
+
+	backoff := wait.Backoff{
```
```suggestion
	deploymentScaleToZeroBackOff := wait.Backoff{
```

fabriziopandini commented on test/e2e/config/docker.yaml:
```diff
@@ -176,9 +176,10 @@ variables:
-  INIT_WITH_BINARY: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v0.4.3/clusterctl-{OS}-{ARCH}"
-  INIT_WITH_PROVIDERS_CONTRACT: "v1alpha4"
-  INIT_WITH_KUBERNETES_VERSION: "v1.22.0"
+  # FIXME: revert this before merge. This has just been added to make it testable on the PR
+  INIT_WITH_BINARY: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v0.3.23/clusterctl-{OS}-{ARCH}"
+  INIT_WITH_PROVIDERS_CONTRACT: "v1alpha3"
+  INIT_WITH_KUBERNETES_VERSION: "v1.21.2"
```
I guess this is temporary, right?

sbueringer replied:
Yes (see the comment in the file :))

sbueringer commented:
/test pull-cluster-api-e2e-full-main

/cherrypick release-0.4
(Same problem on relesae-0.4: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-0.4#capi-e2e-mink8s-release-0-4)

sbueringer commented:
Another successful run

/test pull-cluster-api-e2e-full-main

sbueringer commented:
@fabriziopandini I fixed your findings and added a bit more error wrapping and a log line, PTAL.

I would run the test a few more times in the meantime. When you think the PR is ready/okay I would squash the commits and drop the changes to the docker.yaml file. I think it would be good to merge this (and the corresponding release-0.4 cherry-pick) ASAP to get a lot more test runs in CI to see if this change is enough to make the upgrade completely stable.

sbueringer commented:
/test pull-cluster-api-e2e-full-main

fabriziopandini commented:
/lgtm
/approve

fabriziopandini commented:
/hold

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/hold => contains docker.yaml again to verify the fix

sbueringer commented:
First run green, next one:
/test pull-cluster-api-e2e-full-main

fabriziopandini commented on cmd/clusterctl/client/cluster/upgrader.go:
```diff
@@ -390,9 +418,90 @@ func (u *providerUpgrader) doUpgrade(upgradePlan *UpgradePlan) error {
+		return err
+	}
+
+	// Fetch all Deployments belonging to a provider.
+	deploymentList := &appsv1.DeploymentList{}
+	if err := cs.List(ctx,
+		deploymentList,
+		client.InNamespace(provider.Namespace),
+		client.MatchingLabels{
+			clusterctlv1.ClusterctlLabelName: "",
+			clusterv1.ProviderLabelName:      provider.ManifestLabel(),
+		}); err != nil {
+		return errors.Wrapf(err, "failed to list Deployments for provider %s", provider.Name)
+	}
```
What about deployments in capi-webhook (that exists only when starting point is v1alpha3)

sbueringer replied:
I’m not sure if we should also scale them down
As they are only passive webhooks they don’t really lead to any problems 

So as long as the CRDs are not yet patched they will ensure that conversions for the old types still work and after the CRDs are patched they are just unused

fabriziopandini commented:
/lgtm
/approve 

so we can get more signal from periodic jobs

fabriziopandini commented:
/lgtm

sbueringer commented:
/hold cancel
